File: USACO/Training Pages/Section 3.3/Camelot.cpp
/*
ID: tmwilliamlin168
LANG: C++
TASK: camelot
*/

#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define ld long double

ifstream fin ("camelot.in");
ofstream fout ("camelot.out");

struct Pos {
	int i, j;
	Pos() {}
	Pos(int i, int j) {
		this->i=i;
		this->j=j;
	}
};

struct PosDist {
	Pos p;
	int dist;
	PosDist() {}
	PosDist(Pos p, int dist) {
		this->p=p;
		this->dist=dist;
	}
	bool operator<(const PosDist& b) const {
		return dist>b.dist;
	}
};

const int INF=(int)1e6;
const int knightNeighbors[8][2] = {{-2, -1}, {-1, -2}, {2, -1}, {-1, 2}, {-2, 1}, {1, -2}, {2, 1}, {1, 2}};

int main(){
	ios_base::sync_with_stdio(false);
	fin.tie(NULL);
	int r, c, kingI, kingJ, knights=0;
	fin >> r >> c;
	string tempS;
	fin >> tempS;
	kingJ=tempS[0]-'A';
	fin >> kingI;
	--kingI;
	bool hasKnight[30][26]={};
	while(fin >> tempS) {
		if(tempS=="-")
			break;
		int i;
		fin >> i;
		--i;
		hasKnight[i][tempS[0]-'A']=true;
		++knights;
	}
	if(knights==0) {
		fout << "0\n";
		return 0;
	}
	int dist1[30][26][30][26], dist2[30][26][30][26];
	for(int i=0; i<r; ++i) {
		for(int j=0; j<c; ++j) {
			for(int k=0; k<r; ++k) {
				for(int l=0; l<c; ++l) {
					dist1[i][j][k][l]=INF;
					dist2[i][j][k][l]=INF;
				}
			}
		}
	}

	Pos qu[780];
	int qt, qh;
	priority_queue<PosDist> pq;
	for(int i=0; i<r; ++i) {
		for(int j=0; j<c; ++j) {
			qt=0;
			qh=0;
			Pos start(i, j);
			qu[qt++]=start;
			dist1[i][j][i][j]=0;
			while(qh<qt) {
				Pos p=qu[qh++];
				for(int nI=0; nI<8; ++nI) {
					int newI=p.i+knightNeighbors[nI][0], newJ=p.j+knightNeighbors[nI][1];
					if(newI>=0&&newI<r&&newJ>=0&&newJ<c&&dist1[i][j][newI][newJ]>=INF) {
						qu[qt++]=Pos(newI, newJ);
						dist1[i][j][newI][newJ]=dist1[i][j][p.i][p.j]+1;
					}
				}
			}
			for(int k=0; k<r; ++k)
				for(int l=0; l<c; ++l)
					if(dist1[i][j][k][l]<INF)
						pq.push(PosDist(Pos(k, l), dist1[i][j][k][l]+max(abs(k-kingI), abs(l-kingJ))));
			while(!pq.empty()) {
				PosDist pd = pq.top();
				pq.pop();
				if(dist2[i][j][pd.p.i][pd.p.j]<INF)
					continue;
				dist2[i][j][pd.p.i][pd.p.j]=pd.dist;
				for(int nI=0; nI<8; ++nI) {
					int newI=pd.p.i+knightNeighbors[nI][0], newJ=pd.p.j+knightNeighbors[nI][1];
					if(newI>=0&&newI<r&&newJ>=0&&newJ<c&&dist2[i][j][newI][newJ]>=INF)
						pq.push(PosDist(Pos(newI, newJ), dist2[i][j][pd.p.i][pd.p.j]+1));
				}
			}
		}
	}
	int sumDist[30][26];
	for(int i=0; i<r; ++i) {
		for(int j=0; j<c; ++j) {
			if(!hasKnight[i][j])
				continue;
			for(int k=0; k<r; ++k)
				for(int l=0; l<c; ++l)
					sumDist[k][l]+=dist1[i][j][k][l];
		}
	}
	int minMoves=INF;
	for(int i=0; i<r; ++i) {
		for(int j=0; j<c; ++j) {
			if(!hasKnight[i][j])
				continue;
			for(int k=0; k<r; ++k)
				for(int l=0; l<c; ++l)
					minMoves=min(sumDist[k][l]-dist1[i][j][k][l]+dist2[i][j][k][l], minMoves);
		}
	}
	fout << minMoves << '\n';
}

File: USACO/Training Pages1/Section 3.3/Camelot.java
package section_3_3;
/*
ID: tmwilliamlin168
LANG: JAVA
TASK: camelot
*/
import java.io.*;
import java.util.*;

public class camelot {
	static boolean stdin = true;
	static Reader in;
	static PrintWriter out;
	static String filename = "camelot";
	
	static class Solver {
		static int INF=(int)1e6;
		static int[][] knightNeighbors = {{-2, -1}, {-1, -2}, {2, -1}, {-1, 2}, {-2, 1}, {1, -2}, {2, 1}, {1, 2}};
		
		Solver() {
			int r=in.nextInt(), c=in.nextInt(), kingJ=in.next().charAt(0)-'A', kingI=in.nextInt()-1, knights=0;
			boolean[][] hasKnight = new boolean[r][c];
			while(true) {
				String s = in.next();
				if(s==null||s.charAt(0)=='-')
					break;
				hasKnight[in.nextInt()-1][s.charAt(0)-'A']=true;
				++knights;
			}
			if(knights==0) {
				out.println(0);
				return;
			}
			int[][][][] dist1 = new int[r][c][r][c];
			for(int i=0; i<r; ++i)
				for(int j=0; j<c; ++j)
					for(int k=0; k<r; ++k)
						Arrays.fill(dist1[i][j][k], INF);
			Pos[] qu = new Pos[780];
			int qt, qh;
			for(int i=0; i<r; ++i) {
				for(int j=0; j<c; ++j) {
					qt=0;
					qh=0;
					qu[qt++]=new Pos(i, j);
					dist1[i][j][i][j]=0;
					while(qh<qt) {
						Pos p = qu[qh++];
						for(int[] neighbor : knightNeighbors) {
							int newI=p.i+neighbor[0], newJ=p.j+neighbor[1];
							if(newI>=0&&newI<r&&newJ>=0&&newJ<c&&dist1[i][j][newI][newJ]>=INF) {
								qu[qt++]=new Pos(newI, newJ);
								dist1[i][j][newI][newJ]=dist1[i][j][p.i][p.j]+1;
							}
						}
					}
				}
			}
			int[][] sumDist = new int[r][c];
			for(int i=0; i<r; ++i) {
				for(int j=0; j<c; ++j) {
					if(!hasKnight[i][j])
						continue;
					for(int k=0; k<r; ++k)
						for(int l=0; l<c; ++l)
							sumDist[k][l]+=dist1[i][j][k][l];
				}
			}
			int minMoves=INF;
			for(int i=0; i<r; ++i) {
				for(int j=0; j<c; ++j) {
					int c1Moves=0;
					for(int k=0; k<r; ++k)
						for(int l=0; l<c; ++l)
							if(hasKnight[k][l])
								c1Moves+=dist1[i][j][k][l];
					for(int k=0; k<r; ++k) {
						for(int l=0; l<c; ++l) {
							if(!hasKnight[k][l])
								continue;
							int c2Moves=c1Moves-dist1[i][j][k][l];
							for(int m=Math.max(0, kingI-2); m<=Math.min(r-1, kingI+2); ++m)
								for(int n=Math.max(0, kingJ-2); n<=Math.min(c-1, kingJ+2); ++n)
									minMoves=Math.min(c2Moves+dist1[i][j][m][n]+dist1[m][n][k][l]+Math.max(Math.abs(kingI-m), Math.abs(kingJ-n)), minMoves);
						}
					}
				}
			}
			out.println(minMoves);
		}
		
		class Pos {
			int i, j;
			Pos(int i, int j) {
				this.i=i;
				this.j=j;
			}
		}
		
		class PosDist implements Comparable<PosDist> {
			Pos p;
			int dist;
			PosDist(Pos p, int dist) {
				this.p=p;
				this.dist=dist;
			}
			public int compareTo(PosDist o) {
				return dist-o.dist;
			}
		}
	}
	
	static void preprocess() {
		
	}
	
	public static void main(String[] args) throws Exception {
		if(stdin) {
			in = new Reader();
			out = new PrintWriter(System.out);
		} else {
			in = new Reader(filename+".in");
			out = new PrintWriter(filename+".out");
		}
		preprocess();
		for(int testCases=1; testCases>0; --testCases)
			new Solver();
		out.close();
	}
	
	static class Reader {
		BufferedReader br;
		StringTokenizer st;
		Reader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}
		Reader(String filename) throws Exception {
			br = new BufferedReader(new FileReader(filename));
		}
		String next() {
			while(st==null||!st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch(Exception e) {
					return null;
				}
			}
			return st.nextToken();
		}
		int nextInt() {
			return Integer.parseInt(next());
		}
	}
}